# Tibia Auction Market Analysis - Price Prediction Model

Building on the data exploration phase, this notebook focuses on creating a predictive model to estimate character auction prices. Using machine learning techniques, I aim to build a reliable price prediction system that can help players and traders assess character value.

## Analysis Goals:
1. **Correlation Analysis** - identifying key price drivers
2. **Feature Engineering** - creating vocation-specific variables  
3. **Model Development** - building and training ML algorithms
4. **Model Evaluation** - testing accuracy and feature importance

## Research Questions:
- Which character attributes have the strongest correlation with price?
- How accurately can we predict auction prices?
- What's the relative importance of different features?
- How do vocation-specific skills impact pricing?

## Expected Outcomes:
- Quantified relationships between character stats and prices
- Trained model capable of price prediction
- Feature importance ranking for pricing decisions


**Libraries imports**

In [2]:
# data manipulation and analysis
import pandas as pd
import numpy as np

# DB connection
import psycopg2
from sqlalchemy import create_engine, text
import os
from dotenv import load_dotenv

# Visualization
import matplotlib.pyplot as plt
import seaborn as sns

# for ML
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.preprocessing import StandardScaler, LabelEncoder

# statistical analysis
from scipy.stats import pearsonr

print("Libraries imported successfully.")

Libraries imported successfully.


**Connection with database**

In [3]:
import os
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Database connection config from environment variables
DB_CONFIG = {
    'host': os.getenv('DB_HOST', 'localhost'),
    'port': os.getenv('DB_PORT', '5432'),
    'database': os.getenv('DB_NAME', 'auction_data'),
    'user': os.getenv('DB_USER', 'scraper'),
    'password': os.getenv('DB_PASSWORD')
}

# Create connection string
connection_string = f"postgresql://{DB_CONFIG['user']}:{DB_CONFIG['password']}@{DB_CONFIG['host']}:{DB_CONFIG['port']}/{DB_CONFIG['database']}"

# Create engine
engine = create_engine(connection_string)

print("Database connection configured.")


Database connection configured.


In [5]:
# Indexes were created in data_exploration.ipynb

schema_query = """
SELECT 
    a.id AS auction_id,
    a.current_bid,
    a.auction_end,
    c.id AS character_id, 
    c.vocation_id, 
    c.level, 
    c.sex,
    c.achievement_points,
    c.boss_points,
    c.charm_total,
    c.charm_expansion,
    c.prey_slot,
    c.hunting_slot,
    c.transfer,
    c.gems_greater,
    c.outfits_count,
    c.mounts_count,
    c.store_mounts_count,
    c.store_outfits_count,
    c.hirelings_count,
    s.magic, s.axe, s.sword, s.club, s.distance, s.shielding, s.fist,
    w.pvp_type,
    w.battleye,
    w.location
FROM auctions a 
JOIN characters c ON a.character_id = c.id
LEFT JOIN skills s ON c.skills_id = s.id
LEFT JOIN worlds w ON c.world_id = w.id
WHERE a.has_been_bidded = true
    AND a.is_historical = true
"""
auctions_df = pd.read_sql(schema_query, engine)

auctions_df['auction_end_dt'] = pd.to_datetime(auctions_df['auction_end'], unit='s')
auctions_df['auction_month'] = auctions_df['auction_end_dt'].dt.month
auctions_df['auction_day_of_week'] = auctions_df['auction_end_dt'].dt.dayofweek

display(auctions_df.head())
print(auctions_df.info())

auction_id  current_bid  auction_end  character_id  vocation_id  level  \
0     1969938         5002   1759500000       1969938            1    508   
1     1967415         2000   1759136400       1967415            2    346   
2     1919028          268   1753434000       1919028            1    199   
3     1311129         1772   1690966800       1311129            4    326   
4      907809          101   1656639900        907809            1     53   

     sex  achievement_points  boss_points  charm_total  ...   club  distance  \
0  False                 931         5425         6565  ...  80.36     13.88   
1  False                 443         2610         2982  ...  12.94    115.01   
2  False                  45           10          240  ...  41.12     26.42   
3  False                 212           20         1045  ...  12.96     12.64   
4   True                  19            0            5  ...  40.42     24.61   

   shielding   fist    pvp_type  battleye  location      auction_end_dt  \
0     112.91  21.49    Optional     False        BR 2025-10-03 14:00:00   
1     103.26  16.17        Open     False        NA 2025-09-29 09:00:00   
2     104.31  10.24    Optional     False        EU 2025-07-25 09:00:00   
3      37.11  13.63        Open     False        EU 2023-08-02 09:00:00   
4      98.18  12.32  Retro Open     False        NA 2022-07-01 01:45:00   

   auction_month  auction_day_of_week  
0             10                    4  
1              9                    0  
2              7                    4  
3              8                    2  
4              7                    4  

[5 rows x 33 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 815152 entries, 0 to 815151
Data columns (total 33 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   auction_id           815152 non-null  int64         
 1   current_bid          815152 non-null  int64         
 2   auction_end          815152 non-null  int64         
 3   character_id         815152 non-null  int64         
 4   vocation_id          815152 non-null  int64         
 5   level                815152 non-null  int64         
 6   sex                  815152 non-null  bool          
 7   achievement_points   815152 non-null  int64         
 8   boss_points          815152 non-null  int64         
 9   charm_total          815152 non-null  int64         
 10  charm_expansion      815152 non-null  bool          
 11  prey_slot            815152 non-null  bool          
 12  hunting_slot         815152 non-null  bool          
 13  transfer      

## Feature engineering - Vocation-Specific Skills

In [20]:
# 1. Knight main melee skill (max of sword, axe, club)
auctions_df['knight_melee'] = auctions_df[['sword', 'axe', 'club']].max(axis=1)

# 2. Identify dominant weapon type for Knights
auctions_df['weapon_type'] = auctions_df[['sword', 'axe', 'club']].idxmax(axis=1)

# 3. Create vocation-specific main skills
auctions_df['main_skill'] = 0.0

# Knight (vocation_id = 1): melee combat skills
knight_mask = auctions_df['vocation_id'] == 1
auctions_df.loc[knight_mask, 'main_skill'] = auctions_df.loc[knight_mask, 'knight_melee']

# Paladin (vocation_id = 2): distance fighting
paladin_mask = auctions_df['vocation_id'] == 2
auctions_df.loc[paladin_mask, 'main_skill'] = auctions_df.loc[paladin_mask, 'distance']

# Druid (vocation_id = 3): magic level
druid_mask = auctions_df['vocation_id'] == 3
auctions_df.loc[druid_mask, 'main_skill'] = auctions_df.loc[druid_mask, 'magic']

# Sorcerer (vocation_id = 4): magic level
sorcerer_mask = auctions_df['vocation_id'] == 4
auctions_df.loc[sorcerer_mask, 'main_skill'] = auctions_df.loc[sorcerer_mask, 'magic']

# Monk (vocation_id = 5): fist fighting
monk_mask = auctions_df['vocation_id'] == 5
auctions_df.loc[monk_mask, 'main_skill'] = auctions_df.loc[monk_mask, 'fist']

# shielding is important for Knights and Paladins, so its needed to add secondary skill, to model know, that for them its important
auctions_df['secondary_skill'] = 0.0
auctions_df.loc[auctions_df['vocation_id'].isin([1, 2]), 'secondary_skill'] = auctions_df['shielding']

# 4.  percentile ranking for main_skill within each vocation
auctions_df['main_skill_percentile'] = auctions_df.groupby('vocation_id')['main_skill'].rank(pct=True)

print("Created features:")
print("knight_melee: max(sword, axe, club) for all characters")
print("weapon_type: dominant weapon type")
print("main_skill: vocation-appropriate primary skill")
print("main_skill_percentile: normalized main skill within vocation")

# Display average main skills by vocation
vocation_mapping = {
    1: 'Knight',
    2: 'Paladin', 
    3: 'Druid',
    4: 'Sorcerer',
    5: 'Monk'
}

print("\n-- Main Skill Analysis by Vocation --")
for vocation_id in sorted(auctions_df['vocation_id'].unique()):
    if vocation_id in vocation_mapping:
        vocation_name = vocation_mapping[vocation_id]
        v_data = auctions_df[auctions_df['vocation_id'] == vocation_id]
        
        # Obliczamy więcej statystyk
        stats = v_data['main_skill'].describe()
        
        print(f"\n {vocation_name.upper()}:")
        print(f"Count:  {int(stats['count'])} characters")
        print(f"Average: {stats['mean']:.2f}")
        print(f"Median:  {stats['50%']:.2f}")
        print(f"Max:     {stats['max']:.2f}")
        if vocation_id == 1:
            weapon_dist = v_data['weapon_type'].value_counts(normalize=True) * 100
            print("  - Weapon Meta (Popularity):")
            for w_type, percent in weapon_dist.items():
                print(f"{w_type.capitalize()}: {percent:.1f}%")
        if vocation_id == 2: # Paladin
            print(f"  - Avg Shielding: {v_data['shielding'].mean():.2f}")


Created features:
knight_melee: max(sword, axe, club) for all characters
weapon_type: dominant weapon type
main_skill: vocation-appropriate primary skill
main_skill_percentile: normalized main skill within vocation

-- Main Skill Analysis by Vocation --

 KNIGHT:
Count:  249867 characters
Average: 108.12
Median:  109.88
Max:     144.18
  - Weapon Meta (Popularity):
Sword: 48.0%
Axe: 29.3%
Club: 22.7%

 PALADIN:
Count:  220784 characters
Average: 111.85
Median:  113.50
Max:     145.34
  - Avg Shielding: 99.71

 DRUID:
Count:  149761 characters
Average: 85.05
Median:  84.59
Max:     147.48

 SORCERER:
Count:  186499 characters
Average: 88.27
Median:  88.67
Max:     142.59

 MONK:
Count:  6878 characters
Average: 88.72
Median:  90.70
Max:     122.03


**Conclusions**
**Skills gap**: The data clearly shows that "numbers aren't equal." Knights and Paladins usually have skills around 110–113, while Mages sit around 85–88. This is normal because Magic Level is much harder to train. Your use of percentiles is a smart move—it tells the model that an 88 ML Druid is just as "elite" as a 110 Axe Knight.

**The "Sword" Meta**: Almost half of all Knights (48%) use Swords. This means Sword characters are the most common on the market. Clubs are the rarest (22.7%). This distribution helps the model understand if certain weapon types carry a "rarity premium" or if they are cheaper because they are less popular.

**Shielding matters**: Paladins have a high average shielding (99.71). By adding secondary_skill, you’re helping the model distinguish between a "rushed" character (high distance, low shield) and a "well-rounded" one, which usually sells for more.

**Top-tier Outliers**: There are some characters with skills up to 144-147. These are the "unicorns" of the market. They don't follow normal price rules and will likely be the most expensive auctions in this dataset.

**Monks**: There's only about 6,800 Monks. Since they aren't a standard profession in the main game, these are likely special collectors' items or from a specific game mode. Keeping them separate or labeled is good because their prices won't follow the same logic as a level 500 Sorcerer.